galley/pkg/config/meshcfg/inmemory.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package meshcfg
+
+import (
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+
+	"istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/galley/pkg/config/event"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// InMemorySource is an event.InMemorySource implementation for meshconfig. When the mesh config is first set, add & fullsync events
+// will be published. Otherwise a reset event will be sent.
+type InMemorySource struct {
+	mu      sync.Mutex
+	current *v1alpha1.MeshConfig
+
+	handlers event.Handler
+
+	synced  bool
+	started bool
+}
+
+var _ event.Source = &InMemorySource{}
+
+// NewInmemory returns a new meshconfig.InMemorySource.
+func NewInmemory() *InMemorySource {
+	return &InMemorySource{
+		current: Default(),
+	}
+}
+
+// Dispatch implements event.Dispatcher
+func (s *InMemorySource) Dispatch(handler event.Handler) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.handlers = event.CombineHandlers(s.handlers, handler)
+}
+
+// Start implements event.InMemorySource
+func (s *InMemorySource) Start() {
+	s.mu.Lock()
+
+	if s.started {
+		// Already started
+		s.mu.Unlock()
+		return
+	}
+	s.started = true
+
+	if s.synced {
+		go s.sendInitialEvents()
+	} else {
+		s.mu.Unlock()
+	}
+}
+
+func (s *InMemorySource) sendInitialEvents() {
+	// Lock must be held when entering this go routine
+	s.send(event.Added)
+	s.send(event.FullSync)
+	s.mu.Unlock()
```

Comment:
This is pretty ugly and may not be robust (to release a lock from one go routine which was acquired by a different go routine).

Could you instead use a WaitGroup? The function above then inits the WG, acquires the lock, waits on the WG, and then release the lock. The goroutine merely sends the two events and decrements the wG.

Reply:
I don't think this works. The problem is deadlock avoidance, where the go-routine calling into start/stop can actually hold locks that needs to be acquired by the event handling code. The go-routine needs to be released (i.e. it cannot block), but we also need to publish event before releasing the lock, so that the semantics are preserved.

Reply:
ok, I experimented with a simpler model which now works with soon-to-come-code. Simplified. PTAL.